inngest-core/src/main/kotlin/io/inngest/Step.kt

```diff
@@ -69,16 +69,20 @@ class Step(val state: State) {
     inline fun <reified T> run(id: String, fn: () -> T): T {
         val hashedId = state.getHashFromId(id)
 
-        // TODO - Catch Step Error here and throw it when error parsing is added to getState
-        val stepResult = state.getState<T>(hashedId)
-
-        // If there is no existing result, run the action
-        if (stepResult == null) {
+        try {
+            val stepResult = state.getState<T>(hashedId)
+            if (stepResult is T) {
+                return stepResult
+            }
+        } catch (e: StateNotFound) {
+            // If there is no existing result, run the lambda
             val data = fn()
             throw StepInterruptException(id, hashedId, data)
         }
+        // TODO - Catch Step Error here and throw it when error parsing is added to getState
 
-        return stepResult;
+        // TODO - handle invalidly stored step types properly
```

Comment:
(No action here - just context for posterity)

As we discussed, @djfarrelly, currently the SDK is the piece of the puzzle that decides what operation this state is for, which means that changing a step's type but keeping the same name can produce weird results, memoizing state into steps that don't make sense.

Now that we nest (and as we version opcodes away from nesting), we can include the `op` in the memoized response, ensuring that the step is the same.

Incidentally, the `v0` execution version would hash the entire op which would include the ID and opcode, circumventing this problem. This had other problems and keeping IDs simple feels good; nesting was a good choice!